src/rust/engine/fs/store/src/remote.rs

```diff
@@ -363,12 +363,13 @@ impl ByteStore {
     };
     let result_future = async move {
       let store2 = store.clone();
-      let mut client = store2.cas_client.as_ref().clone();
-      let request = request.clone();
-      let response = client
-        .find_missing_blobs(request)
-        .map_err(status_to_str)
-        .await?;
+      let client = store2.cas_client.as_ref().clone();
+      let response = retry_call(client, move |mut client| {
+        let request = request.clone();
```

Comment:
I do not like having to `clone` the request here, but it is required by Rust's ownership rules since the Tonic client consumes the request. Are there alternatives to this?

Reply:
Ah, I found the issue! If you were to change `retry_call` to be `FnOnce`, then you can avoid the clone. But of course, it needs to be `Fn` because you call the function multiple times.

```diff
diff --git a/src/rust/engine/fs/store/src/remote.rs b/src/rust/engine/fs/store/src/remote.rs
index 43e5fbecb..d68ab59dc 100644
--- a/src/rust/engine/fs/store/src/remote.rs
+++ b/src/rust/engine/fs/store/src/remote.rs
@@ -365,7 +365,7 @@ impl ByteStore {
       let store2 = store.clone();
       let client = store2.cas_client.as_ref().clone();
       let response = retry_call(client, move |mut client| {
-        let request = request.clone();
+        // let request = request.clone();
         async move { client.find_missing_blobs(request).await }
       })
       .await
diff --git a/src/rust/engine/grpc_util/src/retry.rs b/src/rust/engine/grpc_util/src/retry.rs
index 60f817b83..8680be2f5 100644
--- a/src/rust/engine/grpc_util/src/retry.rs
+++ b/src/rust/engine/grpc_util/src/retry.rs
@@ -24,16 +24,17 @@ fn is_retryable(status: &Status) -> bool {
 pub async fn retry_call<T, C, F, Fut>(client: C, f: F) -> Result<Response<T>, Status>
 where
   C: Clone,
-  F: Fn(C) -> Fut,
+  F: FnOnce(C) -> Fut,
   Fut: Future<Output = Result<Response<T>, Status>>,
 {
   const INTERVAL_DURATION: Duration = Duration::from_millis(10);
   const MAX_RETRIES: u32 = 3;
   const MAX_BACKOFF_DURATION: Duration = Duration::from_secs(5);
 
-  let mut last_error: Option<Status> = None;
+  let last_error: Option<Status> = None;
 
   let mut num_retries = 0;
+  f(client);
   while num_retries < MAX_RETRIES {
     // Delay before the next send attempt if this is a retry.
     if num_retries > 0 {
@@ -43,18 +44,18 @@ where
       tokio::time::sleep(sleep_time).await;
     }
 
-    let client2 = client.clone();
-    let result_fut = f(client2);
-    match result_fut.await {
-      Ok(r) => return Ok(r),
-      Err(status) => {
-        if is_retryable(&status) {
-          last_error = Some(status);
-        } else {
-          return Err(status);
-        }
-      }
-    }
+    // let client2 = client.clone();
+    // let result_fut = f(client2);
+    // match result_fut.await {
+    //   Ok(r) => return Ok(r),
+    //   Err(status) => {
+    //     if is_retryable(&status) {
+    //       last_error = Some(status);
+    //     } else {
+    //       return Err(status);
+    //     }
+    //   }
+    // }
 
     num_retries += 1
   }
```

From `rustc --explain E0507`:

> This can also happen when using a type implementing `Fn` or `FnMut`, as neither
allows moving out of them (they usually represent closures which can be called
more than once). Much of the text following applies equally well to non-`FnOnce`
closure bodies.

Reply:
Yeah, because tonic takes ownership of the request, there's no improvement to be had here. In theory I can imagine `tonic`/`prost` allowing you to encode from an `Arc<Message>` or similar, but they don't, so... :) On the plus side, a request shouldn't be _too_ expensive to clone, right?

The clone of `client`, on the other hand, could theoretically be avoided... I think this really wants to be a `FnMut() -> Fut` rather than a `Fn(C) -> Fut` ("Can be called multiple times, but never more than once at a time"), but async closures are unstable and also kinda funky so I don't know of a way to _actually_ avoid the clone of `client`.

Reply:
> The clone of client, on the other hand, could theoretically be avoided... I think this really wants to be a FnMut() -> Fut rather than a Fn(C) -> Fut ("Can be called multiple times, but never more than once at a time"), but async closures are unstable and also kinda funky so I don't know of a way to actually avoid the clone of client.

The client at least is intended to be cheap to Clone (via an Arc).